File: ios/TerminalDeck/App/SessionShelf.swift
/**
 * Which sessions this phone keeps out of the way, and which it keeps at the top.
 *
 * Asad, on the phone's session list: *"swipe left/right on a session row should
 * reveal buttons, WhatsApp-style — close the session (with a confirmation),
 * archive, move. When we will have a lot of sessions we will not like to have
 * all of them over here."* And, about what the gesture does today: *"swipe
 * currently just opens the session, which tapping already does. It's nonsense to
 * keep this feature. We need to use it for something important."*
 *
 * The sentence that decides what this module is, is the second half of the
 * first quote. The problem being described is **a list that is too long to read**,
 * not a machine with too many processes on it — so the fix is a property of this
 * phone's list rather than a command sent to the machine, and both verbs here
 * change what is drawn and nothing else.
 *
 * ## Archive means archive, in the sense the word already has
 *
 * Every phone this app is used from already has this word on it, and it means one
 * thing on all of them: the row leaves the list, the thing itself carries on, and
 * it is one tap to get back. That is exactly what this does. A session that is
 * archived is still running on the machine, still listed by the desktop, still
 * producing output, still able to raise an alert — and none of that is hidden,
 * because the screen that lists archived rows says so in a sentence.
 *
 * The alternative reading — archive meaning *stop it and put it away* — is a
 * thing this app cannot do at all today, and doing half of it would be worse
 * than doing none. See `SessionListView`'s note on Close.
 *
 * ## Pin is the "move"
 *
 * He asked for a third action called move, in a list where the two others were
 * close and archive. On a phone there is nowhere for a session to move *to*: it
 * lives in a folder on a machine, chosen when it was started, and neither of
 * those can change afterwards. What a person moving a row in a list of forty
 * actually wants is the row **at the top**, which is what every list on this
 * phone calls pinning, and it is the other half of the same complaint — archive
 * pushes the noise down, pinning pulls the two that matter up.
 *
 * So the word on the button is `Pin`, because that is what it does and what the
 * platform calls it. Naming it `Move` and having it do this would be a control
 * whose label describes something else, which is the single most repeated
 * complaint in the whole review.
 *
 * ## Both are per machine, and both live on the phone
 *
 * A phone paired with a Mac and a Windows PC holds two unrelated lists, and a
 * store keyed on the session id alone would hide a row on one machine because of
 * a swipe on the other. Ids are minted by each machine's own session layer and
 * nothing makes them unique across machines — the same reason `DeckModel.Route`
 * carries a host — so every entry here is keyed by host **and** session.
 *
 * They live on the phone rather than on the desktop because they are about this
 * screen. Somebody who archives forty finished sessions from the sofa has not
 * asked for their Mac's sidebar to change, and a phone that could reorder the
 * desktop's list would be a surprising thing to hand to a guest.
 *
 * ## Nothing is ever forgotten on its own
 *
 * Not on unpairing, and not when a session exits. Both are deliberate and both
 * are the same argument `PortBook` makes about names: a few dozen bytes of dead
 * text costs nothing, and somebody who unpairs a machine by accident should not
 * also lose an afternoon of tidying. What *is* bounded is how much of it can
 * accumulate — see `trim`.
 */

import Foundation
import Observation

/**
 * Deliberately **not** `@MainActor`, for the two reasons `PortBook` gives and in
 * particular the second: a screen holds it as `var shelf: SessionShelf = .shared`,
 * which is a default argument on a memberwise initialiser, and those are
 * evaluated in a non-isolated context where a main-actor `shared` cannot be
 * named at all.
 */
@Observable
final class SessionShelf {

    /// The one the screens read. A property of this phone rather than of a
    /// machine, so it is a singleton beside `PortBook` and `TextSize` rather
    /// than something threaded through `DeckModel`.
    static let shared = SessionShelf()

    /**
     * How many ids are kept per machine, per list.
     *
     * A bound rather than a cleanup rule, because there is no moment at which
     * this store learns that a session is gone for good: the desktop stops
     * listing an id when its pty is reaped, and a phone that treated "not in the
     * last list" as "delete the record" would forget an archive every time the
     * socket dropped mid-refresh. So the record survives and the *store* is
     * bounded instead, oldest first.
     *
     * Two hundred is far past what a person will ever archive by hand on a phone
     * and still nothing on disk — an id is 26 characters. The number exists so
     * that a machine which churns through sessions for a year cannot grow this
     * without limit, not to enforce a workflow.
     */
    static let maxPerHost = 200

    private let defaults: UserDefaults
    private static let storageKey = "terminaldeck.sessionShelf.v1"

    /**
     * host id → the session ids that machine's list should not draw, in the
     * order they were archived.
     *
     * An array rather than a `Set`, and the order is what makes {@link trim}
     * possible: dropping the oldest needs an oldest, and a set has none. Lookup
     * is a linear scan over at most `maxPerHost` short strings, which is nothing
     * next to the work of drawing the row it decides about.
     */
    private var archived: [String: [String]] = [:]

    /// host id → the session ids pinned to the top, most recently pinned first.
    /// Same shape and same bound as the list above; the order is the display
    /// order rather than only a trimming order.
    private var pinned: [String: [String]] = [:]

    /// `defaults` is a seam for the tests, which run against a suite of their own
    /// so a test cannot archive a session on the machine it is running from.
    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        load()
    }

    // MARK: - Archive

    func isArchived(host: String, session: String) -> Bool {
        archived[host]?.contains(session) == true
    }

    /// Every archived id on one machine, for the screen that lists them.
    func archived(host: String) -> [String] {
        archived[host] ?? []
    }

    /**
     * Put a row away, or bring it back.
     *
     * Archiving also **unpins**, because the two states contradict each other:
     * a row cannot be both at the top of the list and absent from it, and
     * leaving a stale pin behind would make an unarchive jump the row to the top
     * for reasons nobody could see. Unarchiving does not re-pin — a row that
     * comes back comes back where it belongs.
     */
    func setArchived(_ on: Bool, host: String, session: String) {
        guard !host.isEmpty, !session.isEmpty else { return }
        if on {
            setPinned(false, host: host, session: session, save: false)
            var ids = archived[host] ?? []
            guard !ids.contains(session) else { return }
            ids.append(session)
            archived[host] = Self.trim(ids)
        } else {
            archived[host]?.removeAll { $0 == session }
            if archived[host]?.isEmpty == true { archived.removeValue(forKey: host) }
        }
        save()
    }

    // MARK: - Pin

    func isPinned(host: String, session: String) -> Bool {
        pinned[host]?.contains(session) == true
    }

    func setPinned(_ on: Bool, host: String, session: String) {
        setPinned(on, host: host, session: session, save: true)
    }

    /**
     * Pinning also **unarchives**, the mirror of the rule above and for the same
     * reason: pinning something is a statement that it matters, and the only
     * honest response to that is to put it back on the screen.
     *
     * The private `save` flag exists so that the two verbs can call each other
     * without writing the store twice for one gesture. It is not an optimisation
     * — a double write is cheap — it is so that a reader of `setArchived` sees
     * one save at the end of it and knows there is exactly one commit point.
     */
    private func setPinned(_ on: Bool, host: String, session: String, save shouldSave: Bool) {
        guard !host.isEmpty, !session.isEmpty else { return }
        if on {
            archived[host]?.removeAll { $0 == session }
            if archived[host]?.isEmpty == true { archived.removeValue(forKey: host) }
            var ids = pinned[host] ?? []
            ids.removeAll { $0 == session }
            // Newest first: the row somebody just pinned goes to the very top,
            // which is where they are looking when they let go of the swipe.
            ids.insert(session, at: 0)
            pinned[host] = Array(ids.prefix(Self.maxPerHost))
        } else {
            pinned[host]?.removeAll { $0 == session }
            if pinned[host]?.isEmpty == true { pinned.removeValue(forKey: host) }
        }
        if shouldSave { save() }
    }

    // MARK: - The order a list is drawn in

    /**
     * Split one machine's sessions into what the list draws and what it hides.
     *
     * A pure function over values, and separate from the store's own state for
     * the reason `PortCatalog` is separate from `PortBook`: every rule about
     * ordering can then be pinned by a test with no simulator, no host and no
     * `UserDefaults`, and the view has one call rather than three predicates it
     * could get subtly out of step.
     *
     * The pinned rows keep **the order they were pinned in**, not the order the
     * machine sent. That is the whole point of the gesture: a person who pins two
     * sessions is stating which one they want first, and re-sorting them by
     * whatever the desktop's list order happens to be would throw that away. The
     * rest keep the machine's order exactly, because that order is `HostLink`'s
     * and this has no business having an opinion about it.
     */
    func split(_ sessions: [RemoteSession], host: String) -> (listed: [RemoteSession], archived: [RemoteSession]) {
        guard !host.isEmpty else { return (sessions, []) }
        let hidden = archived[host] ?? []
        let order = pinned[host] ?? []
        guard !hidden.isEmpty || !order.isEmpty else { return (sessions, []) }

        var listed: [RemoteSession] = []
        var away: [RemoteSession] = []
        var top: [RemoteSession] = []
        for session in sessions {
            if hidden.contains(session.id) {
                away.append(session)
            } else if order.contains(session.id) {
                top.append(session)
            } else {
                listed.append(session)
            }
        }
        // Sorted by where the id sits in the pin list rather than by the
        // machine's order. `firstIndex` cannot be nil here — every member of
        // `top` was put there by the `contains` above — and the fallback keeps
        // this total rather than trapping if that ever stops being true.
        top.sort { (order.firstIndex(of: $0.id) ?? 0) < (order.firstIndex(of: $1.id) ?? 0) }
        return (top + listed, away)
    }

    /// How many of one machine's *current* sessions are archived. The count on
    /// screen has to be the number of rows the screen behind it will draw, so it
    /// is measured against the live list rather than against the store — a
    /// machine that has been rebooted has archived ids for sessions that no
    /// longer exist, and a badge counting those is a badge that opens onto
    /// nothing.
    func archivedCount(_ sessions: [RemoteSession], host: String) -> Int {
        split(sessions, host: host).archived.count
    }

    // MARK: - Storage

    /// Oldest first, so the drop is of the thing archived longest ago. See
    /// `maxPerHost` for why there is a bound at all.
    private static func trim(_ ids: [String]) -> [String] {
        ids.count <= maxPerHost ? ids : Array(ids.suffix(maxPerHost))
    }

    private struct Stored: Codable {
        var archived: [String: [String]]
        var pinned: [String: [String]]
    }

    private func load() {
        guard let data = defaults.data(forKey: Self.storageKey),
              let stored = try? JSONDecoder().decode(Stored.self, from: data) else { return }
        // Bounded on the way back out as well as on the way in, exactly as
        // `PortBook` cleans its names on load: a record written by another build
        // — or edited by hand in a simulator — must not be able to get around
        // the limit this store promises to hold.
        archived = stored.archived.mapValues { Self.trim($0) }
        pinned = stored.pinned.mapValues { Array($0.prefix(Self.maxPerHost)) }
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(Stored(archived: archived, pinned: pinned)) else { return }
        defaults.set(data, forKey: Self.storageKey)
    }
}

File: android/app/src/main/java/dev/terminaldeck/android/store/SessionShelf.kt
package dev.terminaldeck.android.store

import android.content.Context
import dev.terminaldeck.android.protocol.RemoteSessionView
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Which of a machine's sessions this phone has put away, and which it has pulled to the top.
 *
 * Asad, asking for the swipe: *"close the session (with a confirmation), archive, move. When we will
 * have a lot of sessions we will not like to have all of them over here."* This is the *over here* —
 * the place a row goes, and therefore the place it comes back from. A transcription of
 * `ios/TerminalDeck/App/SessionShelf.swift`.
 *
 * ## It is about this phone, not about the machine
 *
 * Nothing here reaches the wire and nothing here stops a session. Archiving is a **view** on a list
 * this device is shown; the agent on the other machine carries on exactly as it was. That is the one
 * thing about this feature that can be dangerously misread, which is why the screen that lists the
 * archived rows says it in a sentence rather than leaving it to be inferred.
 *
 * ## Pure over a store, so the ordering rules can be tested
 *
 * [split] is a function of values with no Android in it, for the reason the wire codecs are: every
 * rule about ordering can be pinned by a unit test with no device, and the list screen makes one
 * call rather than holding three predicates that could drift out of step.
 */
class SessionShelf(private val store: Store) {

    /** Where the shelf is kept. A seam so a test can drive one that is not a device's disk. */
    interface Store {
        fun read(): String?
        fun write(value: String)
    }

    /**
     * host id → the session ids that machine's list should not draw, in the order they were archived.
     *
     * A list rather than a set, and the order is what makes [trim] possible: dropping the oldest
     * needs an oldest, and a set has none. Lookup is a linear scan over at most [MAX_PER_HOST] short
     * strings, which is nothing next to the work of drawing the row it decides about.
     */
    private val archived = HashMap<String, List<String>>()

    /** host id → the session ids pinned to the top, most recently pinned first. */
    private val pinned = HashMap<String, List<String>>()

    init {
        load()
    }

    fun isArchived(host: String, session: String): Boolean = archived[host]?.contains(session) == true

    /** Every archived id on one machine, for the screen that lists them. */
    fun archived(host: String): List<String> = archived[host].orEmpty()

    /**
     * Put a row away, or bring it back.
     *
     * Archiving also **unpins**, because the two states contradict each other: a row cannot be both
     * at the top of the list and absent from it, and leaving a stale pin behind would make an
     * unarchive jump the row to the top for reasons nobody could see. Unarchiving does not re-pin — a
     * row that comes back comes back where it belongs.
     */
    fun setArchived(on: Boolean, host: String, session: String) {
        if (host.isEmpty() || session.isEmpty()) return
        if (on) {
            setPinned(false, host, session, save = false)
            val ids = archived[host].orEmpty()
            if (session in ids) return
            archived[host] = trim(ids + session)
        } else {
            val ids = archived[host].orEmpty().filterNot { it == session }
            if (ids.isEmpty()) archived.remove(host) else archived[host] = ids
        }
        save()
    }

    fun isPinned(host: String, session: String): Boolean = pinned[host]?.contains(session) == true

    fun setPinned(on: Boolean, host: String, session: String) = setPinned(on, host, session, save = true)

    /**
     * Pinning also **unarchives**, the mirror of the rule above and for the same reason: pinning
     * something is a statement that it matters, and the only honest response to that is to put it
     * back on the screen.
     *
     * The private save flag exists so the two verbs can call each other without writing the store
     * twice for one gesture. Not an optimisation — a double write is cheap — but so that a reader of
     * [setArchived] sees one save at the end of it and knows there is exactly one commit point.
     */
    private fun setPinned(on: Boolean, host: String, session: String, save: Boolean) {
        if (host.isEmpty() || session.isEmpty()) return
        if (on) {
            val left = archived[host].orEmpty().filterNot { it == session }
            if (left.isEmpty()) archived.remove(host) else archived[host] = left
            // Newest first: the row somebody just pinned goes to the very top, which is where they
            // are looking when they let go of the swipe.
            val ids = listOf(session) + pinned[host].orEmpty().filterNot { it == session }
            pinned[host] = ids.take(MAX_PER_HOST)
        } else {
            val ids = pinned[host].orEmpty().filterNot { it == session }
            if (ids.isEmpty()) pinned.remove(host) else pinned[host] = ids
        }
        if (save) save()
    }

    /** A machine that was forgotten. Its shelf goes with it rather than waiting for a re-pair. */
    fun forget(host: String) {
        if (archived.remove(host) == null && pinned.remove(host) == null) return
        save()
    }

    /**
     * Split one machine's sessions into what the list draws and what it hides.
     *
     * The pinned rows keep **the order they were pinned in**, not the order the machine sent. That is
     * the whole point of the gesture: a person who pins two sessions is stating which one they want
     * first, and re-sorting them by whatever the machine's list order happens to be would throw that
     * away. The rest keep the machine's order exactly, because that order is the machine's and this
     * has no business having an opinion about it.
     */
    fun split(sessions: List<RemoteSessionView>, host: String): Split {
        if (host.isEmpty()) return Split(sessions, emptyList())
        val hidden = archived[host].orEmpty()
        val order = pinned[host].orEmpty()
        if (hidden.isEmpty() && order.isEmpty()) return Split(sessions, emptyList())

        val listed = mutableListOf<RemoteSessionView>()
        val away = mutableListOf<RemoteSessionView>()
        val top = mutableListOf<RemoteSessionView>()
        for (session in sessions) {
            when {
                session.id in hidden -> away += session
                session.id in order -> top += session
                else -> listed += session
            }
        }
        // Sorted by where the id sits in the pin list rather than by the machine's order. The index
        // cannot be missing — every member of `top` was put there by the check above — and the
        // fallback keeps this total rather than throwing if that ever stops being true.
        top.sortBy { order.indexOf(it.id).takeIf { index -> index >= 0 } ?: 0 }
        return Split(top + listed, away)
    }

    /**
     * How many of one machine's *current* sessions are archived.
     *
     * Measured against the live list rather than against the store: a machine that has been rebooted
     * has archived ids for sessions that no longer exist, and a count of those is a menu item that
     * opens onto nothing.
     */
    fun archivedCount(sessions: List<RemoteSessionView>, host: String): Int =
        split(sessions, host).archived.size

    data class Split(val listed: List<RemoteSessionView>, val archived: List<RemoteSessionView>)

    /* ------------------------------------------------------------------- storage -- */

    @Serializable
    internal data class Stored(
        val archived: Map<String, List<String>> = emptyMap(),
        val pinned: Map<String, List<String>> = emptyMap(),
    )

    private fun load() {
        val raw = store.read() ?: return
        val stored = try {
            JSON.decodeFromString(Stored.serializer(), raw)
        } catch (_: Exception) {
            // A record written by another build, or half-written when the process died. A shelf that
            // refused to load would be an app that shows an archived session again on every launch.
            return
        }
        // Bounded on the way back out as well as on the way in: a record written by another build
        // must not be able to get around the limit this store promises to hold.
        archived.clear()
        pinned.clear()
        for ((host, ids) in stored.archived) archived[host] = trim(ids)
        for ((host, ids) in stored.pinned) pinned[host] = ids.take(MAX_PER_HOST)
    }

    private fun save() {
        store.write(JSON.encodeToString(Stored.serializer(), Stored(archived.toMap(), pinned.toMap())))
    }

    companion object {
        /**
         * The most rows one machine's shelf may hold.
         *
         * A bound rather than none, because this is written on every swipe and read on every launch:
         * an unbounded list of ids for sessions that stopped existing months ago is a file that only
         * grows. Two hundred is far past any real list and far below anything that costs a frame.
         */
        const val MAX_PER_HOST = 200

        private val JSON = Json { ignoreUnknownKeys = true }

        private fun trim(ids: List<String>): List<String> =
            if (ids.size <= MAX_PER_HOST) ids else ids.takeLast(MAX_PER_HOST)

        private const val FILE = "terminaldeck.preferences"
        private const val KEY = "terminaldeck.sessionShelf.v1"

        /** The real store: the same preferences file the appearance and the alert switches use. */
        fun on(context: Context): SessionShelf = SessionShelf(object : Store {
            private val prefs =
                context.applicationContext.getSharedPreferences(FILE, Context.MODE_PRIVATE)

            override fun read(): String? = prefs.getString(KEY, null)

            override fun write(value: String) {
                prefs.edit().putString(KEY, value).apply()
            }
        })
    }
}
